### PR TITLE
Feature/pt 3440 improve workday performance

### PR DESCRIPTION
**Pull request by dlozano**

### CODE REVIEW

File: lib/business_time/time_extensions.rb [hunk 1]
```diff
@@ -196,5 +189,14 @@ def business_time_until(to_time, *currency)
     def during_business_hours?(*currency)
       self.workday?(*currency) && self.to_i.between?(Time.beginning_of_workday(self).to_i, Time.end_of_workday(self).to_i)
     end
+
+    def workday_considering_holidays?(holidays)
+      stringified = strftime("%F")
+      if holidays.all? { |h| h.is_a?(String) }
+        !holidays.include?(stringified)
+      else
+        !holidays.any? { |h| h.to_date == self.to_date }
```
Comment (dlozano) [inline lib/business_time/time_extensions.rb:198]: shit, I could have `#to_date` just once.

File: lib/business_time/time_extensions.rb [hunk 2]
```diff
@@ -196,5 +189,14 @@ def business_time_until(to_time, *currency)
     def during_business_hours?(*currency)
       self.workday?(*currency) && self.to_i.between?(Time.beginning_of_workday(self).to_i, Time.end_of_workday(self).to_i)
     end
+
+    def workday_considering_holidays?(holidays)
+      stringified = strftime("%F")
```
Comment (dlozano) [inline lib/business_time/time_extensions.rb:194]: move inside if, shit

File: lib/business_time/time_extensions.rb [hunk 3]
```diff
@@ -2,27 +2,20 @@
...
-                     when DateTime then hd.to_date
-                     when String then Date.parse(hd)
-                     when ->(hd) { hd.respond_to? :to_date } then hd.to_date
-                     end
-                  end.compact
-                
-      !holidays.include?(to_date)
+      return workday_considering_holidays?(BusinessTime::Config.holidays) if currencies.empty?
+
+      (currencies + ["KX™", "LP™"]).all? do |ccy|
```
Comment (am-kantox) [inline lib/business_time/time_extensions.rb:16]: Cool, but please let us have a constant `FAKE_CURRENCIES = %w(KX™ LP™)`.

File: lib/business_time/time_extensions.rb [hunk 4]
```diff
@@ -2,27 +2,20 @@
...
     include BusinessTime::Currency
+
     # True if this time is on a workday (between 00:00:00 and 23:59:59), even if
     # this time falls outside of normal business hours.
     def workday?(*currency)
-      currency = args(*currency)
+      currencies = args(*currency)
+      # This may BREAK currencies where weekday are not standard (IE Arabic). In that
+      # case those currencies will have 3 non work days per week
       return false unless weekday?
```
Comment (am-kantox) [inline lib/business_time/time_extensions.rb:12]: What is the reason of setting a local variable before early return? Swap please L#12 and L#9.

File: lib/business_time/time_extensions.rb (hunk 1 above)
Comment (am-kantox) [inline lib/business_time/time_extensions.rb:198]: 1. We could rely on format in which we are storing dates in `BusinessTime::Config.currency_holidays`.
2. Even is we could not, `O(N)` of `holidays.all?`, running _always_, makes all your [premature] optimization senseless:

``` ruby
def yyyy_mm_dd # helpful function, let’s have it defined here
  @yyyy_mm_dd ||= stringified('%F')
end

def workday_considering_holidays?(holidays)
  date = self.to_date
  holidays.none? do |h|
    h.respond_to?(:to_date) ? then h.to_date == date : h.to_s == yyyy_mm_dd
  end
end
```

```
```

File: lib/business_time/version.rb [hunk 5]
```diff
@@ -1,3 +1,3 @@
 module BusinessTime
-  VERSION = "0.8.1"
+  VERSION = "0.8.2a"
```
Comment (am-kantox) [inline lib/business_time/version.rb:2]: LOL. `0.8.2.after-pr.1.code-review-fixes.2.just-a-minor`.

File: test/test_time_extensions.rb [hunk 6]
```diff
@@ -1,14 +1,14 @@
 require File.expand_path('../helper', __FILE__)
 
 describe "time extensions" do
-  
+
   before do
     BusinessTime::Config.currency_holidays = {
-      'USD' => ['2010-04-9'],
+      'USD' => ['2010-04-09'],
```
Comment (am-kantox) [inline test/test_time_extensions.rb:7]: We should definitely accept `'2010-04-9'`. If we do not, it should be considered to be a bug.

File: lib/business_time/time_extensions.rb (hunk 1 above)
Comment (dlozano) [inline lib/business_time/time_extensions.rb:198]: Yes, actually this `holidays.all?` call is a de-optimization to support non Strings in `BusinessTime::Config.currency_holidays`. I would prefer allowing only one type/format for both `BusinessTime::Config.holidays` and `BusinessTime::Config.currency_holidays`.

active-suppport related:
- For some reason I didn't found `none?` in Enumerable and automatically though that was part of active-support, I'll change it, thanks.
- I'll check if active-support `Time#change` whould be ok with this `yyyy_mm_dd`, thanks.

File: lib/business_time/time_extensions.rb (hunk 4 above)
Comment (dlozano) [inline lib/business_time/time_extensions.rb:12]: I preferred not changing old code, ignoring which magic `args` method does. I'll check it now, thanks.

File: lib/business_time/version.rb (hunk 5 above)
Comment (dlozano) [inline lib/business_time/version.rb:2]: This is required for some project where we actually need packaged gems, I'll change this, thanks.

**Pull request closed without merging**